elastic/apm-server
# Changing TLS client_authentication default and cleaning up docs

## Motivation/summary
This PR covers a couple of changes related to APM Server's SSL settings for communicating with the agents. 

It changes the default value for `apm-server.ssl.client_authentication` from `optional` to `none`. Sending client certificates is not supported by any of the backend agents. It can be used by the RUM agent when a client certificate is configured in the browser. It can also be used by Jaeger agents when connecting via gRPC, which is an experimental feature so far. 
The default being set to `optional` does not prevent anyone from connecting to the server without providing a certificate. If users want to rely on client certification verification the value should be changed to `required`. For more details see https://github.com/elastic/apm-server/issues/3442#issuecomment-600628214. 

The PR also aims to clarify the documentation around the client certificate related configuration options (`apm-server.ssl.client_authentication` and `apm-server.ssl.certificate_authorities`) in the `apm-server.yml` and the docs. See https://github.com/elastic/apm-server/issues/3498 and https://github.com/elastic/apm-server/issues/3494.

This PR also removes the configuration option `apm-server.ssl.verification_mode` from the `apm-server.yml` and the docs. The [`VerificationMode`](https://github.com/elastic/beats/blob/master/libbeat/common/transport/tlscommon/server_config.go#L85) gets internally translated to [`tls.Config InsecureSkipVerify`](https://github.com/elastic/beats/blob/master/libbeat/common/transport/tlscommon/tls_config.go#L99). 
From the [golang/src/crypto/tls/common.go `InsecureSkipVerify`](https://github.com/golang/go/blob/master/src/crypto/tls/common.go#L548):
> InsecureSkipVerify controls whether a client verifies the
> server's certificate chain and host name.
> If InsecureSkipVerify is true, TLS accepts any certificate
> presented by the server and any host name in that certificate.
> In this mode, TLS is susceptible to man-in-the-middle attacks.
> This should be used only for testing.

Configuring the verification mode on the server does not have any effect. 





## Checklist

~- [ ] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).~



## How to test these changes

Start APM Server with TLS and make a request from the browser (Chrome). Ensure no prompt for chosing a certificate appears.

## Related issues
https://github.com/elastic/apm-server/issues/3198
https://github.com/elastic/apm-server/issues/3494
https://github.com/elastic/apm-server/issues/3498
https://github.com/elastic/apm-server/issues/3442
